dynamiqs/core/_utils.py

```diff
@@ -78,26 +78,90 @@ def is_shape(x: object) -> bool:
 
 
-def _flat_vectorize(
-    f: callable, n_batch: PyTree[int], out_axes: PyTree[int | None]
-) -> callable:
-    # todo: write doc
+def tree_false_to_none(
+    tree: PyTree, is_leaf: callable[PyTree, bool] | None = None
+) -> PyTree:
+    """Replace all `False` values in a tree by `None`."""
+    return jtu.tree_map(lambda x: x if x is not False else None, tree, is_leaf=is_leaf)
+
+
+def _flat_vectorize(  # noqa: C901
+    f: TimeArray, n_batch_false: PyTree, out_axes_false: PyTree
+) -> TimeArray:
+    """Vectorize a Hamiltonian function.
+
+    Args:
+        f: the Hamiltonian function.
+        n_batch_false: the batch shape of the Hamiltonian.
+        out_axes_false: the out axes of the Hamiltonian.
+    """
+    # JAX completely dismisses leaves with a `None` when applying the `tree_map`, so we
+    # need to keep one version of the batch shape with `False` instead of `None`
+    # to keep the structure.
+    n_batch = tree_false_to_none(n_batch_false, is_leaf=is_shape)
+    out_axes = tree_false_to_none(out_axes_false)
+
     broadcast_shape = jtu.tree_leaves(n_batch, is_shape)
     broadcast_shape = jnp.broadcast_shapes(*broadcast_shape)
-    in_axes = jtu.tree_map(
-        lambda x: 0 if len(x) > 0 else None, n_batch, is_leaf=is_shape
-    )
-
-    for _ in range(len(broadcast_shape)):
-        f = jax.vmap(f, in_axes=in_axes, out_axes=out_axes)
 
-    return f
+    def tree_map_fn(i: int, x: tuple[int, ...]) -> int | None:
+        """Args:
+            i: the index in `broadcast_shape`.
+            x: the shape of the leaf.
+
+        Returns:
+            The in axes for the vmap function.
+        """
+        if len(x) <= i or x[len(x) - i - 1] == 1:
+            return None
+        else:
+            return 0
+
+    n = len(broadcast_shape)
+    expand_dims = []  # dimensions '1' that will be lost during the vmap but that
+    # we want to keep in the results.
+    for i in range(len(broadcast_shape)):
+        in_axes = jtu.tree_map(ft.partial(tree_map_fn, i), n_batch, is_leaf=is_shape)
+        if jtu.tree_all(jtu.tree_map(lambda x: x is None, in_axes)):
+            expand_dims.append(n - i - 1)
+        else:
+            f = jax.vmap(f, in_axes=in_axes, out_axes=out_axes)
+
+    expand_dims = sorted(expand_dims)
+
+    def squeeze_args(size: PyTree, arg: PyTree) -> PyTree:
+        """Squeeze the all the arguments with a dimension 1."""
+        if is_shape(size):
+            for i, s in reversed(list(enumerate(size))):
+                if s == 1:
+                    arg = arg.squeeze(i)
+
+        return arg
+
+    def unsqueeze_args(out_ax: PyTree, result: PyTree) -> PyTree:
+        """Unsqueeze the result."""
+        if out_ax is not False:
+            for dim in expand_dims:
+                result = jtu.tree_map(
+                    ft.partial(lambda t, dim: jnp.expand_dims(t, dim), dim=dim), result
+                )
+
+        return result
+
+    def wrap(*args: [PyTree]) -> PyTree:
```

Comment:
I think both versions are necessary, but the `def wrap(*args: [PyTree]) -> PyTree:` is sufficient, because it doesn't allow `args` to be a dict for instance

Reply:
The usage of `necessary` and `sufficient` is terrible here, so tell me if it's not clear

Reply:
Sorry I'm not sure what you mean haha. What do you accept with the typing `*args: [PyTree]`? For me, the typing `*args: PyTree` simply means that all elements of the `args` tuple should be PyTrees.

Reply:
Oh right, I see what you mean, I though that all `*args` had to be batched with `[Something]` since it's always an array but it turns out that it's wrong